### ISSUE TITLE
RStudio slow to open R files in directories with many files, especially on network drives

### ISSUE THREAD

**Issue by bapike** [rstudio/rstudio#14096]
I have a network drive containing a single directory with about 2000 files, some of which are small R scripts.  Opening one of these R scripts in RStudio using File->Open File (or the "Open an existing file" button) can take a long time (40 seconds on a LAN, or >20 minutes over a VPN), while opening the same file in Notepad over the same network connection takes 2 seconds.

Note: When a file is opened via File->Recent Files, or when a file opens automatically when RStudio starts (because it was open when RStudio last closed), then the file opens in RStudio within a second or two even over VPN.

After opening a file that demonstrates the problem, using "Help->Diagnostics->Request Log" shows that most of the time is occupied by a `list_files` request for the directory containing the file I'm trying to open.  The response to that request is a list of all the files in that directory, with file metadata.  I'm not sure how this list is useful to RStudio, as it seems unnecessary.  The file seems to be opened only after the `list_files` request completes.

Additionally, this `list_files` request appears to be inefficient.  Running Process Monitor on RStudio while trying to open a problematic file shows a lot of repetition in the requests for information about each file; for example, the attributes for each file seem to be requested about 16 times.  Assuming that the `list_files` request can be run using `.rs.invokeRpc("list_files",mydir,FALSE,FALSE)`, then I've benchmarked `list_files` unfavorably against some similar commands (see below).

I am not using a Project nor using version control from within RStudio.  I turned off as much as possible in my user preferences: R code completion, R diagnostics, checks for missing R packages, spell-checking, version control interface for RStudio projects, and publishing documents/apps/APIs.  I recently updated from RStudio-1.1.442, which may have had a similar problem but perhaps not as severe.

### System details

    RStudio Edition :  RStudio Desktop
    RStudio Version :  RStudio 2023.12.0+369 "Ocean Storm" Release (33206f75bd14d07d84753f965eaa24756eda97b7, 2023-12-17) for windows
    OS Version      :  Windows 10 (22H2)
    R Version       :  R version 4.3.2 (2023-10-31 ucrt)

### Steps to reproduce the problem

1. Create a directory containing lots of files (say, 8000 for a local directory, 1000 for a network drive on a LAN, 30 for a network drive on a VPN, but you may need to adjust).
2. In RStudio, open one of these files using File->Open File.  Time how long it takes to open.
3. In Notepad, open the same file.  Time how long it takes to open.

Here's a script for generating a directory with lots of files:

```
root_directory<-tempdir()
num_files<-8000L

# Create an empty directory
mydir<-file.path(root_directory,"lots_of_files")
unlink(mydir,recursive=TRUE)
stopifnot(dir.exists(mydir)==FALSE)
dir.create(mydir)
stopifnot(dir.exists(mydir)==TRUE)

# Create a file to open
myfile<-file.path(mydir,"script.R")
write.csv(data.frame(),file=myfile)

# Create a bunch of other files
for (i in seq.int(1L,num_files,by=1L)) {
  fname<-file.path(mydir,sprintf("SomeFile%05d.txt",i))
  write.csv(data.frame(),file=fname)
}
cat("Now, time how long to open ",gsub("/","\\\\",myfile),"\n",sep="")

# Cleanup
unlink(mydir,recursive=TRUE)
stopifnot(dir.exists(mydir)==FALSE)
```

### Describe the problem in detail

The file takes much longer to open in RStudio than to open in Notepad.

The time it takes to open the file in RStudio increases as the number of files in its directory increases.

### Describe the behavior you expected

The time it takes to open the file in RStudio should be similar to the time it takes to open the file in Notepad.

The time it takes to open a file in RStudio should not depend on the number of files in its directory.

- [X] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [X] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [x] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.posit.co/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### Benchmarking of variations of `list_files`

Here are some commands similar to `list_files`.  `fs::dir_info` appears to be the fastest.

```
microbenchmark::microbenchmark(
  RStudioRPC=.rs.invokeRpc("list_files",mydir,FALSE,FALSE),
  R1=file.info(list.files(mydir,full.names=TRUE,include.dirs = FALSE,recursive=FALSE)),
  R2=lapply(list.files(mydir,full.names=TRUE,include.dirs = FALSE,recursive=FALSE),file.info),
  R3=fs::dir_info(mydir),
  times=10L,
  control=list(order="random",warmup=2L)
)
```

**Comment by ronblum** [comment]
@bapike Thank you for raising this! Do you see a difference in behavior depending on whether or not you're using native dialogs? You can switch between types using <kbd>Global Options -> General -> Advanced -> Use native file and message dialog boxes</kbd>

<img width="270" alt="Screenshot 2024-01-08 at 10 43 50" src="https://github.com/rstudio/rstudio/assets/31009952/e43ada2c-d437-45a7-a03f-17d9f8174ddc">

**Comment by bapike** [comment]
The `Use native file and message dialog boxes` option seems to produce no difference in the time it takes to load a particular file:
`Use native file and message dialog boxes` checked: 43 seconds
`Use native file and message dialog boxes` unchecked: 42 seconds
`Use native file and message dialog boxes` checked: 41.5 seconds
`Use native file and message dialog boxes` unchecked: 42 seconds

(starting RStudio fresh each time and immediately loading the file)

**Comment by kevinushey** [comment]
> After opening a file that demonstrates the problem, using "Help->Diagnostics->Request Log" shows that most of the time is occupied by a list_files request for the directory containing the file I'm trying to open. The response to that request is a list of all the files in that directory, with file metadata. I'm not sure how this list is useful to RStudio, as it seems unnecessary. The file seems to be opened only after the list_files request completes.

Doesn't File -> Open File present you with a file viewer for the files within that directory? That is, the list_files RPC is made so that RStudio can populate the view of available files in the directory you're trying to use; it's not done in response to actually opening the file.

Here's what I see in the Request Log after using File -> Open File to open a file in the project:

<img width="273" alt="Screenshot 2024-01-09 at 10 43 00 AM" src="https://github.com/rstudio/rstudio/assets/1976582/f5440d18-d3d8-4a42-8893-03ebfed7279d">

Note the initial list_files RPC is made to populate the file list; after selecting a file it's later opened via the open_document RPC (and then a bunch of other stuff happens).

What if you open the file from R using `file.edit("path/to/script.R")`; is that similarly slow?

**Comment by bapike** [comment]
My earlier timing was done by pasting a full file path into the file picker.  The default directory is my user profile "Documents" directory (the same as `getwd()`, and which RStudio is calling `~`), which doesn't have a lot of files.  The files I'm having the most trouble with are of the form `network-drive:\script.R`, but I also see delays in opening `network-drive:\subdirectory\script.R`.

With `native file and message dialog boxes` enabled:
* `File->Open File`->(see listing of Documents directory)->`Cancel` produces no `list_files` event.
* `File->Open File`->Browse to `network-drive:\` (no delay at this step) -> `Cancel` produces no `list_files` event.  
* `File->Open File`->Browse to `network-drive:\` (no delay at this step) -> select a file -> `Open`-> file opens after ~40 seconds.  There is one `list_files` event.
* `File->Open File`->pasting full path to file->`Open` -> file opens after ~40 seconds. There is one `list_files` event.

With `native file and message dialog boxes` disabled:
* `File->Open File`->(see listing of Documents directory)->`Cancel` produces a `list_files` event for `~`.
* `File->Open File`->(see listing of Documents directory)-> Pasting `network-drive:/`-> `Open` -> ~40 second delay with swirling icon -> the list of files is still that for Documents -> `Cancel`, shows 3 `list_files` events: one for `~`, one for `network-drive:/`, and then one for `~`.  So, the file picker does not let me browse.  Weird.
* `File->Open File`->(see listing of Documents directory)-> Pasting `network-drive:/subdirectory` -> `Open` -> Swirling icon and some delay -> list of files for `network-drive:/subdirectory` -> Selecting `..` -> `Open` -> ~40 second delay and list of files for `network-drive:/` -> `Cancel`, shows 4 `list_files` events: one for `~`, one for `network-drive:/`, one for `network-drive:/subdirectory`, and one for `network-drive:/`.
* `File->Open File`->(see listing of Documents directory)-> Pasting `network-drive:/subdirectory` -> `Open` -> Swirling icon and some delay -> list of files for `network-drive:/subdirectory` -> Selecting `..` -> `Open` -> ~40 second delay and list of files for `network-drive:/` -> selecting filename -> `Open`, opens the file quickly.  There are 4 `list_files` events:  one for `~`, one for `network-drive:/`, one for `network-drive:/subdirectory`, and one for `network-drive:/`.

With the non-native file picker I can't see a way to browse outside subdirectories of `~` without typing in an alternate directory.

Using `file.edit("path/to/script.R")` opens the file immediately, which is a great workaround to my problem.  Thank you for pointing that out!

Here's a screenshot of the Request Log when using the native file picker, pasting a full path, and opening a file:
![image](https://github.com/rstudio/rstudio/assets/7726997/7a2eefb6-ae87-43ef-8e04-80565372be9a)

**Comment by kevinushey** [comment]
Thanks! The extra context here helps; I think I now understand the issue and can now reproduce locally.

The offending code is here:

https://github.com/rstudio/rstudio/blob/5f8c81f4623f50975a8a1dd8553404e2c1777428/src/gwt/src/org/rstudio/studio/client/common/impl/DesktopFileDialogs.java#L199

and I think the intention here is to ensure that the working directory is saved, so that if you use File -> Open multiple times in a session, the last-used directory is preserved. I'll try to look a bit more.

**Comment by kevinushey** [comment]
@bapike: we have a potential fix for this issue in the latest daily builds available at https://dailies.rstudio.com; would you be able to test?

**Comment by bapike** [comment]
I'm testing this build:
```RStudio 2024.04.0-daily+297 "Chocolate Cosmos" Daily (8a7fa575187f3ed04e2a4d75147b85e177bb248e, 2024-01-25) for windows```

With `native file and message dialog boxes` enabled, the issue has been fixed (**Thank you!**):
* `File->Open File`->Browse to `network-drive:\` (no delay at this step) -> select a file -> `Open`-> file opens immediately. There are zero `list_files` events.
* `File->Open File`->pasting full path to file->`Open` -> file opens immediately.  There are zero `list_files` events.


With `native file and message dialog boxes` disabled, the issue is still present:
* `File->Open File`->(see listing of Documents directory)-> Pasting `network-drive:/`-> `Open` -> ~40 second delay with swirling icon -> the list of files is still that for Documents -> `Cancel`, shows 3 `list_files` events: one for `~`, one for `network-drive:/`, and then one for `~`.  So, the file picker does not let me browse.
* `File->Open File`->(see listing of Documents directory)-> Pasting `network-drive:/subdirectory` -> `Open` -> Swirling icon and some delay -> list of files for `network-drive:/subdirectory` -> Selecting `..` -> `Open` -> ~40 second delay and list of files for `network-drive:/` -> `Cancel`, shows 4 `list_files` events: one for `~`, one for `network-drive:/`, one for `network-drive:/subdirectory`, and one for `network-drive:/`.
* `File->Open File`->(see listing of Documents directory)-> Pasting `network-drive:/subdirectory` -> `Open` -> Swirling icon and some delay -> list of files for `network-drive:/subdirectory` -> Selecting `..` -> `Open` -> ~40 second delay and list of files for `network-drive:/` -> selecting filename -> `Open`, opens the file quickly.  There are 4 `list_files` events:  one for `~`, one for `network-drive:/`, one for `network-drive:/subdirectory`, and one for `network-drive:/`.
* `File->Open File`->(see listing of Documents directory)-> Pasting `network-drive:/filename.R` -> `Open` -> Swirling icon and ~40 second delay -> file opens.  There are 2 `list_files` events:  one for `~`, and one for `network-drive:/`.

I see three _possible_ remaining issues: A) when a full path is provided to the non-native file picker, an unnecessary file listing is done, B) the file listing in the non-native file picker is slow when there are many files in a directory (probably because `list_files` is slow), and C) the non-native file picker does not allow one to open a root directory like `drive:/` (local or network).

Since I use the native file picker, the issue no longer affects me.  I don't know how many users use the non-native option, so I'm not sure how important A, B, and C are.  I'll let you decide whether to close this issue.

**Comment by ronblum** [comment]
@kevinushey Moving back to "Ready" due to @bapike's with native dialogs disabled.

**Comment by kevinushey** [comment]
I think we can mark this issue as closed given that the specific issue @bapike was encountering is resolved. We can reconsider the behavior with non-native dialogs in the future if other users report having trouble?

At least, using `file.edit()` directly from the R console is a way of side-stepping this particular behavior, so there are workarounds for any users affected by this.

(Putting back into test so @ronblum can make the final judgment call!)

**Comment by ronblum** [comment]
Works for me!

**Comment by victordebuen** [comment]
Hi
I think I have the same problem but with RStudio Server and I can't use RStudio Desktop because it's a remote machine without graphical environment. I would like to know if there is any patch or test release in the form of a deb package that I could install to test it.

Thanks in advance!

**Comment by ronblum** [comment]
@victordebuen What version of RStudio Server and R are you using, and what operating system? And are the files on the same drive that RStudio Server is running on, or are they on a remote drive?

**Comment by victordebuen** [comment]
Hi @ronblum 

All the files and the PostgreSQL database are on the same server.
 - OS: Ubuntu 22.04.3 LTS, jammy
 - R version 4.1.2 (2021-11-01) -- ‘Bird Hippie’.
 - RStudio Server 2023.09.1 Build 494

I access that server via SSH for console runnable tasks but use RStudio Server via VPN for development and anything that involves data visualisation on the fly.

**Comment by ronblum** [comment]
@victordebuen Can you open a new issue for this, and include the details above and describe what causes the problem? It seems like it might be a different setup, and I don't want the issue to get lost here. Thanks!
